### PR TITLE
Use NNPACK for strided convolutions.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Use NNPACK for strided convolutions.

### CODE REVIEW

File: aten/src/ATen/native/NNPACK.cpp [hunk 1]
```diff
@@ -232,76 +236,75 @@ Tensor _nnpack_spatial_convolution(
...
+    };
+
+    const auto input_ = input.contiguous();
+    // If we don't have a defined bias Tensor, we need to create one filled with zeroes
+    const auto bias_ = bias.defined() ? bias : at::zeros({weight.size(0)}, input.options());
+
+    const size_t input_size_per_batch = input_channels * input_size.width * input_size.height;
+    const size_t output_size_per_batch = output_channels * output_size.width * output_size.height;
+
+    for (size_t batch = 0u; batch < batch_size; ++batch) {
```
Comment (ljk53) [inline aten/src/ATen/native/NNPACK.cpp]: I didn't trace into the details but looks like nnp_convolution_output() might be able to do better batching than for-loop, or it's equivalent?
Guess it's not a big deal for mobile use case as batch size is usually small (1)?

File: aten/src/ATen/native/NNPACK.cpp [hunk 2]
```diff
@@ -54,6 +55,7 @@ bool _nnpack_available() {
 #else
 
 #include "nnpack.h"
+#include "nnpack/reference.h"
```
Comment (ljk53) [inline aten/src/ATen/native/NNPACK.cpp]: not used?

Comment (AshkanAliabadi) [reply aten/src/ATen/native/NNPACK.cpp]: Right, good catch.

File: aten/src/ATen/native/NNPACK.cpp (hunk 1 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/NNPACK.cpp]: Good point, I'll modify.

**Review by ljk53 — APPROVED** [review]
Thanks!

**Comment by suo** [comment]
This PR broke backcompat tests :( : https://app.circleci.com/jobs/github/pytorch/pytorch/3549012

**Comment by suo** [comment]
@houseroad what's the best thing to do here? Revert? Whitelist this op?

**Comment by houseroad** [comment]
Let's first unloand it to make sure we either make it backward compatible, or no existing serialized models used this op.

**Comment by AshkanAliabadi** [comment]
This PR gives a 20-ish% performance boost for ResNet50 on mobile.  NNPACK does not provide a backwards path for strided convolutions so this is the only way I am aware of that this can be implemented so backwards compatibility may not be possible.  How should I check this is not used in any serialized model?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1965,7 +1965,7 @@
 - func: _nnpack_available() -> bool
   use_c10_dispatcher: full
 
-- func: _nnpack_spatial_convolution(Tensor input, Tensor weight, Tensor? bias, int[2] padding) -> Tensor
+- func: _nnpack_spatial_convolution(Tensor input, Tensor weight, Tensor? bias, int[2] stride, int[2] padding) -> Tensor
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml]: you are changing the signature in a backward incompatible way. 

Can you change it in the following way?
```
_nnpack_spatial_convolution(Tensor input, Tensor weight, Tensor? bias, int[2] padding,  int[2]? stride=None) -> Tensor
```
The new API is backward compatible with the existing API.

**Review by houseroad — COMMENTED** [review]
Sure, I am not saying this PR is not important. My point is that breaking CI is not good. Especially, IMO, there is a way to make the new API to be compatible with existing API. If we break the BC CI, 1) exiting models may break, 2) others may be confused/blocked by the noise from the broken CI.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/native_functions.yaml]: Definitely.  Thanks for your feedback.  Just to be sure, should I unland this CR first, and update this PR to contain that fix and resubmit, or, create a new patch that contains your proposed fix only and apply that to master?

Comment (houseroad) [reply aten/src/ATen/native/native_functions.yaml]: I have already unlanded the PR and fixed the BC CI.

Please resubmit the PR and make the new aten function backward compatible. If you need any help on BC, feel free to ping me.

**Pull request closed without merging**